### PR TITLE
Expand the checks used to determine the RebootPending property value.

### PR DESCRIPTION
**Pull request by barnson**
Add some helper functions to DUtil to help with that.

### CODE REVIEW

File: src/burn/engine/variable.cpp [hunk 1]
```diff
@@ -1961,6 +1964,40 @@ static HRESULT InitializeVariableRebootPending(
         }
     }
 
+    hr = RegKeyReadNumber(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\ServerManager", L"CurrentRebootAttempts", TRUE, &dwValue);
+    fRebootPending |= SUCCEEDED(hr) && 0 < dwValue;
+
+    hr = RegKeyReadNumber(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\Updates", L"UpdateExeVolatile", TRUE, &dwValue);
+    fRebootPending |= SUCCEEDED(hr) && 0 < dwValue;
+
+    hr = RegOpen(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Component Based Servicing\\RebootPending", KEY_READ | KEY_WOW64_64KEY, &hk);
```
Comment (Mike-GC) [inline src/burn/engine/variable.cpp]: Can we use RegValueExists here (for the default value of the key)?

File: src/burn/engine/variable.cpp [hunk 2]
```diff
@@ -1961,6 +1964,40 @@ static HRESULT InitializeVariableRebootPending(
         }
     }
 
+    hr = RegKeyReadNumber(HKEY_LOCAL_MACHINE, L"SOFTWARE\\Microsoft\\ServerManager", L"CurrentRebootAttempts", TRUE, &dwValue);
+    fRebootPending |= SUCCEEDED(hr) && 0 < dwValue;
```
Comment (robmen) [inline src/burn/engine/variable.cpp]: How about a skipping all the tests as soon as one comes back true? No point doing all the checks if first one passes.

File: src/burn/engine/variable.cpp [hunk 3]
```diff
@@ -1945,6 +1945,9 @@ static HRESULT InitializeVariableRebootPending(
     HRESULT hr = S_OK;
     BOOL fRebootPending = FALSE;
     BOOL fComInitialized = FALSE;
+    DWORD dwValue;
+    LPWSTR sczValue = NULL;
```
Comment (Mike-GC) [inline src/burn/engine/variable.cpp]: sczValue is unused

File: src/libs/dutil/regutil.cpp [hunk 4]
```diff
@@ -935,3 +935,59 @@ extern "C" HRESULT DAPI RegQueryKey(
...
+
+/********************************************************************
+RegKeyReadNumber - reads a DWORD registry key value as a number from
+a specified subkey.
+
+*********************************************************************/
+extern "C" HRESULT DAPI RegKeyReadNumber(
+    __in HKEY hk,
+    __in_z LPCWSTR wzSubKey,
+    __in_z_opt LPCWSTR wzName,
```
Comment (Mike-GC) [inline src/libs/dutil/regutil.cpp:947]: I think you want __in_z - assuming NULL would accesses the default value, which I don't think can be a number.

File: src/burn/engine/variable.cpp (hunk 3 above)
Comment (barnson) [inline src/burn/engine/variable.cpp]: fixed

File: src/burn/engine/variable.cpp (hunk 2 above)
Comment (barnson) [inline src/burn/engine/variable.cpp]: Hmmm. True. But man, all those braces...

File: src/libs/dutil/regutil.cpp (hunk 4 above)
Comment (barnson) [inline src/libs/dutil/regutil.cpp:947]: agreed

File: src/burn/engine/variable.cpp (hunk 2 above)
Comment (barnson) [inline src/burn/engine/variable.cpp]: I feel like a shell developer...

File: src/burn/engine/variable.cpp (hunk 1 above)
Comment (barnson) [inline src/burn/engine/variable.cpp]: Let's find out.

File: src/libs/dutil/regutil.cpp (hunk 4 above)
Comment (barnson) [inline src/libs/dutil/regutil.cpp:947]: Turns out it can be. Who knew?

Comment (Mike-GC) [inline src/libs/dutil/regutil.cpp:947]: Learn something new every day.

File: src/burn/engine/variable.cpp [hunk 5]
```diff
@@ -1961,6 +1963,56 @@ static HRESULT InitializeVariableRebootPending(
...
+
+                        if (!fRebootPending)
+                        {
+                            fRebootPending = RegValueExists(HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Session Manager", L"PendingFileRenameOperations", TRUE);
+
+                            if (!fRebootPending)
+                            {
+                                fRebootPending = RegValueExists(HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Session Manager", L"PendingFileRenameOperations2", TRUE);
+
+                                if (!fRebootPending)
```
Comment (Mike-GC) [inline src/burn/engine/variable.cpp:1997]: I'm not a fan of the enormous number of nested if's as Bob said. While they help readers of the code understand the flow a bit better, the incredible amount of indentation hinders readability. Though putting all the huge indentations in a separate function would make it more palatable.

Comment (robmen) [inline src/burn/engine/variable.cpp:1997]: <shrug/> One alternative is to make the if's all flat and check them many times... which would be fine (but not immediately obvious). However, checking a bunch of registry keys when you already know the answer... that's not good. So this change is important.

Comment (barnson) [inline src/burn/engine/variable.cpp:1997]: The approach of so many four-line blocks:

```
if (fRebootPending)
{
    ExitFunction();
}
```

over and over made me die a little inside. I could invent a new ExitOnFalse/ExitOnTrue macro...

**Pull request merged**